### PR TITLE
Add support for tree walking tests ;)

### PR DESCRIPTION
**Pull request by porcelli**
Hi Alan,

 First of all congrats for this nice piece of software ;). Well here is a small contribution that adds tree walking support to this nice tool ;)

Best Regards,
Alexandre Porcelli

### CODE REVIEW

**Comment by maguro** [comment]
Great!  There's a lot of code commented out.  Is there a reason for that?

File: junit/src/main/java/com/toolazydogs/aunit/Work.java [hunk 1]
```diff
@@ -115,49 +132,188 @@ public final class Work
...
+
+        return true;
+    }
+
+
+    public static ArgumentBuilder args(Object... arguments) {
+        return new ArgumentBuilder(arguments);
     }
 
+    public static ArgumentBuilder arg(Object argument) {
```
Comment (maguro) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java]: Why do we need this method?  Does not the method at 211 suffice?

File: junit/src/main/java/com/toolazydogs/aunit/Work.java [hunk 2]
```diff
@@ -115,49 +132,188 @@ public final class Work
...
+        return new ArgumentBuilder(argument);
+    }
+
+    public static TreeBuilder resultOf(Tree tree) {
+        return new TreeBuilder(tree);
+    }
+
+    public static class ArgumentBuilder {
```
Comment (maguro) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java:210]: It seems that the argument builder is syntactic sugar and not really required.  Is this true?  If so what's your opinion on this?

I think that it might make things a bit too wordy.  For example when an expression has a single argument it seems to get wordy.

File: junit/src/main/java/com/toolazydogs/aunit/Work.java [hunk 3]
```diff
@@ -115,49 +132,188 @@ public final class Work
...
+        ArgumentBuilder(Object... arguments){
+            this.arguments = arguments;
+        }
+
+        Object[] get(){
+            return this.arguments;
+        }
+    }
+
+    public static class TreeBuilder {
```
Comment (maguro) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java:221]: Not sure this should be an inner class.  Everything else is "external".  If we adopt this idiom then maybe we should bring more things into the Work class.  Thoughts?

File: junit/src/main/java/com/toolazydogs/aunit/Work.java [hunk 4]
```diff
@@ -115,49 +132,188 @@ public final class Work
...
+        TreeBuilder(Tree tree){
+            this.tree = tree;
+        }
+
+        Tree get(){
+            return this.tree;
+        }
+    }
+
+//    /**
```
Comment (maguro) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java:232]: Not sure why all this is commented out.

File: junit/src/main/java/com/toolazydogs/aunit/internal/TreeParserFactory.java [hunk 5]
```diff
@@ -16,32 +16,161 @@
...
     private final Class<? extends TreeParser> treeParserClass;
+    private final boolean failOnError;
+    private Class<? extends TreeParser> wrapperClass;
 
-    public TreeParserFactory(Class<? extends TreeParser> treeParserClass)
-    {
-        assert treeParserClass != null;
-
+    public TreeParserFactory(Class<? extends TreeParser> treeParserClass, boolean failOnError) {
         this.treeParserClass = treeParserClass;
```
Comment (maguro) [inline junit/src/main/java/com/toolazydogs/aunit/internal/TreeParserFactory.java:43]: Need assert here

File: it/src/test/java/com/toolazydogs/aunit/CMinusTest.java [hunk 6]
```diff
@@ -128,6 +129,20 @@ public class CMinusTest
         }
     }
 
+    @Test
+    public void testTreeTest() throws Exception
+    {
+        try
+        {
+            walk(withRule("testExpr"), resultOf(parse("1 + 2", rule("expression", arg(15)))));
```
Comment (maguro) [inline it/src/test/java/com/toolazydogs/aunit/CMinusTest.java]: Do we really need withRule() and resultOf() ?

walk("testExpr", parse("1 + 2", rule("expression", 15)))

seems just as good.  What are your thoughts on this?

Comment (porcelli) [inline it/src/test/java/com/toolazydogs/aunit/CMinusTest.java]: This is just to improve readability. I don't think that this is an intuitive instruction:

``` java
walk("testExpr", parse("1 + 2", rule("expression", 15)))
```

In fact it was my first local implemantation, but i wasn't satified with result. It becames a bit strange if you need to pass arguments to the `walk` method.

File: junit/src/main/java/com/toolazydogs/aunit/internal/TreeParserFactory.java (hunk 5 above)
Comment (porcelli) [inline junit/src/main/java/com/toolazydogs/aunit/internal/TreeParserFactory.java:43]: ohh yes... don't know why i removed :P.

File: junit/src/main/java/com/toolazydogs/aunit/Work.java (hunk 4 above)
Comment (porcelli) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java:232]: I commented those methods just 'cos i didn't see any use of them ;)

File: junit/src/main/java/com/toolazydogs/aunit/Work.java (hunk 3 above)
Comment (porcelli) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java:221]: It's a commons idiom to use inner class @ internal DSLs. I can't see why create a file to hold those simple types that are really tied to Works class.

File: junit/src/main/java/com/toolazydogs/aunit/Work.java (hunk 2 above)
Comment (porcelli) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java:210]: This is just a way to enforce the use of `arg` and `args` methods.

File: junit/src/main/java/com/toolazydogs/aunit/Work.java (hunk 1 above)
Comment (porcelli) [inline junit/src/main/java/com/toolazydogs/aunit/Work.java]: This is just to improve a readability consistency, it's up to you keep or remove.

**Comment by maguro** [comment]
Thanks!
